### PR TITLE
:bug: Correctly sets DNS imageRepository when custom repositories are used

### LINKED ISSUE #2810: KCP: CoreDNS upgrade with imageRepository containing subpaths doesn't work
**What steps did you take and what happened:**
Try to set `dns.imageRepository` to a repository that has a subpath like `foo.io/upgrade`, the upgrade fails with:
```
E0327 22:36:32.212262       1 controller.go:258] controller-runtime/controller "msg"="Reconciler error" "error"="failed to update CoreDNS deployment: failed to validate CoreDNS: toVersion \"v1.6.5_foo.4\" must be greater than fromVersion \"v1.6.5_foo.4\""  "controller"="kubeadmcontrolplane" "request"={"Namespace":"default","Name":"test-upgrade"}
```

**What did you expect to happen:**
Upgrade to succeed.

/kind bug
/area control-plane
/milestone v0.3.x

### PR DESCRIPTION
**Pull request by gab-satchi**
**Which issue(s) this PR fixes**
Fixes #2810

### CODE REVIEW

File: util/image/image.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+*/
+
+package image
+
+import (
+	"errors"
+	"strings"
+)
+
+//GetRepository returns the repository path given a full container path
```
Comment (vincepri) [inline util/image/image.go]:
```suggestion
// GetRepository returns the repository path given a full container path
```

File: util/image/image.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	imageNameIndex := strings.LastIndex(path, "/")
+	if imageNameIndex == -1 {
+		return "", errors.New("unable to parse image")
+	}
+	remainder := path[imageNameIndex+1:]
+	if tagIndex := strings.IndexRune(remainder, ':'); tagIndex > -1 {
+		return remainder[:tagIndex], nil
+	} else {
+		return remainder, nil
+	}
```
Comment (vincepri) [inline util/image/image.go]: Can probably drop this else

File: util/image/image.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	return path[:imageNameIndex], nil
+}
+
+//GetImage returns the image given a full container path
+func GetImage(path string) (string, error) {
+	imageNameIndex := strings.LastIndex(path, "/")
+	if imageNameIndex == -1 {
+		return "", errors.New("unable to parse image")
+	}
+	remainder := path[imageNameIndex+1:]
```
Comment (vincepri) [inline util/image/image.go]: This could potentially panic

**Comment by vincepri** [comment]
In alternative, if we can't find a good regex with groups, we can use the reference repository we're using today just for validation / normalization, then parse the strings manually where required

**Comment by gab-satchi** [comment]
Agree with the potential of panics. That was going through my head when working on this as well. My regex-fu isn't that good but I think the reference library uses them as well. So might just be able to get it from there.

**Comment by randomvariable** [comment]
There appears to be a regex library for images here https://github.com/docker/distribution/blob/master/reference/regexp.go

**Comment by vincepri** [comment]
For this iteration, I'd suggest to build these higher level functions on top of the reference library, but make them work like we need.

**Comment by gab-satchi** [comment]
@vincepri updated to just use the constant in the file that's used to search for the deployment.

**Comment by vincepri** [comment]
code changes lgtm, should we retitle the PR and tackle the image library separately?

**Comment by gab-satchi** [comment]
/retitle Correctly sets DNS imageRepository when custom repositories are used

**Pull request merged**